File: app/src/main/java/com/ppy/nfcsample/card/reader/CardClient.java
package com.ppy.nfcsample.card.reader;

import com.ppy.nfclib.NfcCardReaderManager;
import com.ppy.nfcsample.card.DefaultCardInfo;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

/**
 * Created by ZP on 2018/1/5.
 */

public class CardClient {

    private List<IReader> readers;
    private IReader.Chain chain;
    private NfcCardReaderManager mNfcCardReaderManager;


    private CardClient(Builder builder) {
        this.readers = builder.readers;
        this.mNfcCardReaderManager = builder.mNfcCardReaderManager;
        chain = new CardReaderChain(this.readers);
    }

    public DefaultCardInfo execute() throws IOException {
        return chain.proceed(mNfcCardReaderManager);
    }


    public static final class Builder {
        private final List<IReader> readers = new ArrayList<>();
        private NfcCardReaderManager mNfcCardReaderManager;

        public Builder() {
        }

        public Builder(CardClient copy) {
            this.readers.addAll(copy.readers);
            this.mNfcCardReaderManager = copy.mNfcCardReaderManager;
        }

        public Builder addReader(IReader reader) {
            readers.add(reader);
            return this;
        }

        public Builder nfcManager(NfcCardReaderManager manager) {
            mNfcCardReaderManager = manager;
            return this;
        }

        public CardClient build() {
            return new CardClient(this);
        }
    }
}

File: app/src/main/java/com/ppy/nfcsample/card/reader/CardClient.kt
package com.ppy.nfcsample.card.reader

import com.ppy.nfclib.NfcManagerCompat
import com.ppy.nfcsample.card.DefaultCardInfo
import java.io.IOException
import java.util.*

/**
 * Created by ZP on 2018/1/5.
 */

class CardClient private constructor(builder: Builder) {

    private val readers: List<IReader>
    private val chain: IReader.Chain
    private var mNfcCardReaderManager: NfcManagerCompat


    init {
        this.readers = builder.readers
        this.mNfcCardReaderManager = builder.mNfcCardReaderManager
        chain = CardReaderChain(this.readers)
    }

    @Throws(IOException::class)
    fun execute(): DefaultCardInfo? {
        return chain.proceed(mNfcCardReaderManager)
    }


    class Builder {
        val readers = ArrayList<IReader>()
        lateinit var mNfcCardReaderManager: NfcManagerCompat

        fun addReader(reader: IReader) = apply {
            readers.add(reader)
        }

        fun nfcManager(manager: NfcManagerCompat) = apply {
            mNfcCardReaderManager = manager
        }

        fun build(): CardClient {
            return CardClient(this)
        }
    }
}
